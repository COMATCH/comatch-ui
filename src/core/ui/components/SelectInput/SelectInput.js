import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import noop from 'lodash/noop';
import { InputLabel } from '../InputLabel';
import { InputError } from '../InputError';

import './SelectInput.scss';

const propTypes = {
    display: PropTypes.oneOf(['block', 'inline']),
    disabled: PropTypes.bool,
    id: PropTypes.string,
    inputError: PropTypes.string,
    label: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node, PropTypes.string]),
    name: PropTypes.string.isRequired,
    onBlur: PropTypes.func,
    onChange: PropTypes.func,
    options: PropTypes.arrayOf(
        PropTypes.shape({
            value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool]),
            label: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool]),
        }),
    ).isRequired,
    required: PropTypes.bool,
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool]),
};

const defaultProps = {
    display: 'block',
    disabled: false,
    id: null,
    inputError: '',
    label: '',
    onBlur: noop,
    onChange: noop,
    required: false,
    value: undefined,
};

const renderOptionsElements = (optionElements) =>
    optionElements.map((optionElement, index) => (
        <option key={index + 1} value={optionElement.value}>
            {optionElement.label}
        </option>
    ));

export const SelectInput = ({
    display,
    disabled,
    id,
    inputError,
    label,
    name,
    onBlur,
    onChange,
    options,
    required,
    value,
}) => {
    const classes = classNames('Input', 'SelectInput', display, {
        disabled,
        'has-error': inputError,
    });

    return (
        <div id={id} className={classes}>
            {label && <InputLabel text={label} required={required} />}
            <select
                className="SelectInput__input"
                disabled={disabled}
                name={name}
                onBlur={onBlur}
                onChange={onChange}
                value={value}
            >
                {renderOptionsElements(options)}
            </select>
            {inputError && <InputError text={inputError} />}
        </div>
    );
};

SelectInput.defaultProps = defaultProps;

SelectInput.propTypes = propTypes;
