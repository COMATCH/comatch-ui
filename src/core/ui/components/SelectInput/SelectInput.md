## SelectInput

prop | type | default | required | description
---- | :----: | :-------: | :--------: | -----------
**disabled** | `Boolean` | `false` | :x: | 
**display** | `Enum('block', 'inline')` | `'block'` | :x: | 
**id** | `String` | `null` | :x: | 
**inputError** | `String` | `''` | :x: | 
**label** | `Union<Array[]<ReactNode>\|ReactNode\|String>` | `''` | :x: | 
**label<1>** | `Array[]<ReactNode>` |  | :x: | 
**label<2>** | `ReactNode` |  | :x: | 
**label<3>** | `String` |  | :x: | 
**name** | `String` |  | :white_check_mark: | 
**onBlur** | `Function` | `noop` | :x: | 
**onChange** | `Function` | `noop` | :x: | 
**options** | `Array[]<Shape>` |  | :white_check_mark: | 
**options[].label** | `Union<String\|Number\|Boolean>` |  | :x: | 
**options[].label<1>** | `String` |  | :x: | 
**options[].label<2>** | `Number` |  | :x: | 
**options[].label<3>** | `Boolean` |  | :x: | 
**options[].value** | `Union<String\|Number\|Boolean>` |  | :x: | 
**options[].value<1>** | `String` |  | :x: | 
**options[].value<2>** | `Number` |  | :x: | 
**options[].value<3>** | `Boolean` |  | :x: | 
**required** | `Boolean` | `false` | :x: | 
**value** | `Union<String\|Number\|Boolean>` | `undefined` | :x: | 
**value<1>** | `String` |  | :x: | 
**value<2>** | `Number` |  | :x: | 
**value<3>** | `Boolean` |  | :x: | 

