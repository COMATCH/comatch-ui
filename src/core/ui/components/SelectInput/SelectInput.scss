@import '../../../styles/variables.scss';

.SelectInput {
    font-size: $co-font-sm;
    display: inline-block;
    margin-bottom: $co-inputs-margin-bottom;

    // common styles
    .InputLabel,
    &.block,
    &__input {
        display: block;
    }

    // Elements
    &__input {
        // coloring
        background-color: $co-color-white;
        border: 1px solid $co-color-mid-gray;
        border-radius: 3px;
        box-shadow: 0 0 4px 0 rgba(1, 156, 185, 0), inset 0 0 2px 0 rgba(193, 193, 193, 0.5);
        // sizing & whitespace
        padding: 8px;
        height: 40px;
        width: 100%;
        // transition effects
        transition: box-shadow 250ms ease-out;
        transform: translateZ(0);
        will-change: box-shadow;
        // reset browser defaults
        outline: 0 !important;

        &:focus {
            box-shadow: 0 0 4px 0 rgba(1, 156, 185, 1), inset 0 0 2px 0 rgba(193, 193, 193, 0.5);
        }
    }

    // States
    &.has-error {
        .SelectInput__input {
            box-shadow: 0 0 4px 0 rgba(173, 25, 25, 1), inset 0 0 2px 0 rgba(193, 193, 193, 0.5);
        }
    }

    &.disabled {
        color: $co-color-gray;
    }
}

// The following styles are used to achive cross browser consistency
.SelectInput__input {
    // This fix will remove the glossy background for select elements
    // from the Safari browser, while maintaing the default arrow icon
    background: url('data:image/svg+xml;base64,PHN2ZyBpZD0iTGF5ZXJfMSIgZGF0YS1uYW1lPSJMYXllciAxIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCA0Ljk1IDEwIj48ZGVmcz48c3R5bGU+LmNscy0xe2ZpbGw6I2ZmZjt9LmNscy0ye2ZpbGw6IzQ0NDt9PC9zdHlsZT48L2RlZnM+PHRpdGxlPmFycm93czwvdGl0bGU+PHJlY3QgY2xhc3M9ImNscy0xIiB3aWR0aD0iNC45NSIgaGVpZ2h0PSIxMCIvPjxwb2x5Z29uIGNsYXNzPSJjbHMtMiIgcG9pbnRzPSIxLjQxIDQuNjcgMi40OCAzLjE4IDMuNTQgNC42NyAxLjQxIDQuNjciLz48cG9seWdvbiBjbGFzcz0iY2xzLTIiIHBvaW50cz0iMy41NCA1LjMzIDIuNDggNi44MiAxLjQxIDUuMzMgMy41NCA1LjMzIi8+PC9zdmc+')
        no-repeat right 2px center;
    // For uniformity across browsers
    -moz-appearance: none;
    -webkit-appearance: none;
    appearance: none;
    padding-right: 20px;

    &:-moz-focusring {
        // This will remove the focus dotted outline appearing
        // only on Firefox achieving cross browser uniformity
        color: transparent;
        text-shadow: 0 0 0 #000;
    }
}
