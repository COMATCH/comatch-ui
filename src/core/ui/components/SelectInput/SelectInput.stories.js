import React from 'react';
import Highlight from 'react-highlight';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { SelectInput } from './SelectInput';

storiesOf('SelectInput', module)
    .add('single with name and options', () => (
        <>
            <SelectInput
                name="fruits"
                options={[
                    { value: 'apple', label: 'Apple' },
                    { value: 'banana', label: 'Banana' },
                    { value: 'peach', label: 'Peach' },
                ]}
            />
            <Highlight className="html">
                {'<SelectInput' +
                    '\n\tname="fruits"' +
                    '\n\toptions={[' +
                    `\n\t\t{ value: 'apple', label: 'Apple' },` +
                    `\n\t\t{ value: 'banana', label: 'Banana' },` +
                    `\n\t\t{ value: 'peach', label: 'Peach' },` +
                    '\n\t]}' +
                    '\n/>'}
            </Highlight>
        </>
    ))
    .add('multiple with name options, display inline and display block', () => (
        <div>
            <SelectInput
                name="fruits"
                options={[
                    { value: 'apple', label: 'Apple' },
                    { value: 'banana', label: 'Banana' },
                    { value: 'peach', label: 'Peach' },
                ]}
            />
            <SelectInput
                display="inline"
                name="fruits"
                options={[
                    { value: 'apple', label: 'Apple' },
                    { value: 'banana', label: 'Banana' },
                    { value: 'peach', label: 'Peach' },
                ]}
            />
            <Highlight className="html">
                {'<SelectInput' +
                    '\n\tname="fruits"' +
                    '\n\toptions={[' +
                    `\n\t\t{ value: 'apple', label: 'Apple' },` +
                    `\n\t\t{ value: 'banana', label: 'Banana' },` +
                    `\n\t\t{ value: 'peach', label: 'Peach' },` +
                    '\n\t]}' +
                    '\n/>\n\n'}
                {'<SelectInput' +
                    '\n\tdisplay="inline"' +
                    '\n\tname="fruits"' +
                    '\n\toptions={[' +
                    `\n\t\t{ value: 'apple', label: 'Apple' },` +
                    `\n\t\t{ value: 'banana', label: 'Banana' },` +
                    `\n\t\t{ value: 'peach', label: 'Peach' },` +
                    '\n\t]}' +
                    '\n/>'}
            </Highlight>
        </div>
    ))
    .add('single with label', () => (
        <div>
            <SelectInput
                label="Select your fruit"
                name="fruits"
                options={[
                    { value: 'apple', label: 'Apple' },
                    { value: 'banana', label: 'Banana' },
                    { value: 'peach', label: 'Peach' },
                ]}
            />
            <Highlight className="html">
                {'<SelectInput' +
                    '\n\tlabel="Select your fruit"' +
                    '\n\tname="fruits"' +
                    '\n\toptions={[' +
                    `\n\t\t{ value: 'apple', label: 'Apple' },` +
                    `\n\t\t{ value: 'banana', label: 'Banana' },` +
                    `\n\t\t{ value: 'peach', label: 'Peach' },` +
                    '\n\t]}' +
                    '\n/>'}
            </Highlight>
        </div>
    ))
    .add('single with label and required', () => (
        <div>
            <SelectInput
                label="Select your fruit"
                name="fruits"
                options={[
                    { value: 'apple', label: 'Apple' },
                    { value: 'banana', label: 'Banana' },
                    { value: 'peach', label: 'Peach' },
                ]}
                required
            />
            <Highlight className="html">
                {'<SelectInput' +
                    '\n\trequired' +
                    '\n\tlabel="Select your fruit"' +
                    '\n\tname="fruits"' +
                    '\n\toptions={[' +
                    `\n\t\t{ value: 'apple', label: 'Apple' },` +
                    `\n\t\t{ value: 'banana', label: 'Banana' },` +
                    `\n\t\t{ value: 'peach', label: 'Peach' },` +
                    '\n\t]}' +
                    '\n/>'}
            </Highlight>
        </div>
    ))
    .add('single with error', () => (
        <div>
            <SelectInput
                inputError="Alarm! Meltdown Imminent!"
                label="Select your fruit"
                name="fruits"
                options={[
                    { value: 'apple', label: 'Apple' },
                    { value: 'banana', label: 'Banana' },
                    { value: 'peach', label: 'Peach' },
                ]}
                required
            />
            <Highlight className="html">
                {'<SelectInput' +
                    '\n\tinputError="Alarm! Meltdown Imminent!"' +
                    '\n\trequired' +
                    '\n\tlabel="Select your fruit"' +
                    '\n\tname="fruits"' +
                    '\n\toptions={[' +
                    `\n\t\t{ value: 'apple', label: 'Apple' },` +
                    `\n\t\t{ value: 'banana', label: 'Banana' },` +
                    `\n\t\t{ value: 'peach', label: 'Peach' },` +
                    '\n\t]}' +
                    '\n/>'}
            </Highlight>
        </div>
    ))
    .add('single with onChange', () => {
        const handleChange = (event) => action('changed')(event.target.name, event.target.value);
        return (
            <>
                <SelectInput
                    name="fruits"
                    onChange={handleChange}
                    options={[
                        { value: 'apple', label: 'Apple' },
                        { value: 'banana', label: 'Banana' },
                        { value: 'peach', label: 'Peach' },
                    ]}
                    required
                />
                <Highlight className="html">
                    {'<SelectInput' +
                        '\n\tonChange={function() {}}' +
                        '\n\trequired' +
                        '\n\tname="fruits"' +
                        '\n\toptions={[' +
                        `\n\t\t{ value: 'apple', label: 'Apple' },` +
                        `\n\t\t{ value: 'banana', label: 'Banana' },` +
                        `\n\t\t{ value: 'peach', label: 'Peach' },` +
                        '\n\t]}' +
                        '\n/>'}
                </Highlight>
            </>
        );
    })
    .add('single with onBlur', () => (
        <div>
            <SelectInput
                name="fruits"
                onBlur={action('blur')}
                options={[
                    { value: 'apple', label: 'Apple' },
                    { value: 'banana', label: 'Banana' },
                    { value: 'peach', label: 'Peach' },
                ]}
                required
            />
            <Highlight className="html">
                {'<SelectInput' +
                    `\n\tonBlur={action('blur')}` +
                    '\n\trequired' +
                    '\n\tname="fruits"' +
                    '\n\toptions={[' +
                    `\n\t\t{ value: 'apple', label: 'Apple' },` +
                    `\n\t\t{ value: 'banana', label: 'Banana' },` +
                    `\n\t\t{ value: 'peach', label: 'Peach' },` +
                    '\n\t]}' +
                    '\n/>'}
            </Highlight>
        </div>
    ))
    .add('with defaultValue and onChange (uncontrolled)', () => {
        const handleChange = (event) => action('changed')(event.target.name, event.target.value);
        return (
            <div>
                <SelectInput
                    name="fruit"
                    onChange={handleChange}
                    options={[
                        { value: 'apple', label: 'Apple' },
                        { value: 'banana', label: 'Banana' },
                        { value: 'peach', label: 'Peach' },
                    ]}
                    defaultValue={'peach'}
                />
                <Highlight className="html">
                    {'<SelectInput' +
                        `\n\tdefaultValue={'peach'}` +
                        '\n\tname="fruits"' +
                        '\n\toptions={[' +
                        `\n\t\t{ value: 'apple', label: 'Apple' },` +
                        `\n\t\t{ value: 'banana', label: 'Banana' },` +
                        `\n\t\t{ value: 'peach', label: 'Peach' },` +
                        '\n\t]}' +
                        '\n/>'}
                </Highlight>
            </div>
        );
    })
    .add('disabled', () => (
        <>
            <SelectInput
                name="fruits"
                disabled
                options={[
                    { value: 'apple', label: 'Apple' },
                    { value: 'banana', label: 'Banana' },
                    { value: 'peach', label: 'Peach' },
                ]}
            />
            <Highlight className="html">
                {'<SelectInput' +
                    '\n\tdisabled' +
                    '\n\tname="fruits"' +
                    '\n\toptions={[' +
                    `\n\t\t{ value: 'apple', label: 'Apple' },` +
                    `\n\t\t{ value: 'banana', label: 'Banana' },` +
                    `\n\t\t{ value: 'peach', label: 'Peach' },` +
                    '\n\t]}' +
                    '\n/>'}
            </Highlight>
        </>
    ))
    .add('as a controlled input', () => (
        <>
            <SelectInputContainer />
            <Highlight className="javascript">
                {`class SelectInputContainer extends React.Component {
    state = { fruit: 'banana' };

    handleChange = (event) => {
        this.setState({ fruit: event.target.value });
        action('changed in container')(event.target.name, event.target.value);
    };

    render() {
        return (
            <SelectInput
                name="fruit"
                options={[
                    { value: 'apple', label: 'Apple' },
                    { value: 'banana', label: 'Banana' },
                    { value: 'peach', label: 'Peach' },
                ]}
                onChange={this.handleChange}
                value={this.state.fruit}
            />
        );
    }
}`}
            </Highlight>
        </>
    ))
    .add('with placeholder', () => (
        <>
            <SelectInput
                name="fruits"
                placeholder="My placeholder"
                options={[
                    { value: 'apple', label: 'Apple' },
                    { value: 'banana', label: 'Banana' },
                    { value: 'peach', label: 'Peach' },
                ]}
            />
            <Highlight className="html">
                {'<SelectInput' +
                    '\n\tplaceholder="My placeholder"' +
                    '\n\tname="fruits"' +
                    '\n\toptions={[' +
                    `\n\t\t{ value: 'apple', label: 'Apple' },` +
                    `\n\t\t{ value: 'banana', label: 'Banana' },` +
                    `\n\t\t{ value: 'peach', label: 'Peach' },` +
                    '\n\t]}' +
                    '\n/>'}
            </Highlight>
        </>
    ));

class SelectInputContainer extends React.Component {
    state = { fruit: 'banana' };

    handleChange = (event) => {
        this.setState({ fruit: event.target.value });
        action('changed in container')(event.target.name, event.target.value);
    };

    render() {
        return (
            <SelectInput
                name="fruit"
                options={[
                    { value: 'apple', label: 'Apple' },
                    { value: 'banana', label: 'Banana' },
                    { value: 'peach', label: 'Peach' },
                ]}
                onChange={this.handleChange}
                value={this.state.fruit}
            />
        );
    }
}
