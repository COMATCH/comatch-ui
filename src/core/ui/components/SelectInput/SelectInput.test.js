import React from 'react';
import renderer from 'react-test-renderer';
import noop from 'lodash/noop';
import { mount } from 'enzyme';
import { SelectInput } from './SelectInput';
import '../../../config/tests/setup';

describe('SelectInput tests', () => {
    const REQUIRED_PROPS = {
        name: 'fruits',
        options: [
            { value: 'apple', label: 'Apple' },
            { value: 'banana', label: 'Banana' },
            { value: 'peach', label: 'Peach' },
        ],
    };

    const OPTIONAL_PROPS = {
        className: 'SelectInput__input',
        disabled: false,
        name: 'fruits',
        onBlur: noop,
        onChange: noop,
        defaultValue: 'peach',
    };

    it('should render and check snapshot correctly', () => {
        const selectInputComponent = renderer.create(<SelectInput {...REQUIRED_PROPS} />).toJSON();
        expect(selectInputComponent).toMatchSnapshot();
    });

    it('should render correctly with optional props', () => {
        const props = {
            ...OPTIONAL_PROPS,
            ...REQUIRED_PROPS,
        };
        const selectInputComponent = mount(<SelectInput {...props} />);
        expect(selectInputComponent.prop('className')).toEqual(props.className);
        expect(selectInputComponent.prop('disabled')).toEqual(props.disabled);
        expect(selectInputComponent.prop('name')).toEqual(props.name);
        expect(selectInputComponent.prop('onBlur')).toEqual(props.onBlur);
        expect(selectInputComponent.prop('onChange')).toEqual(props.onChange);
        expect(selectInputComponent.prop('defaultValue')).toEqual(props.defaultValue);
    });

    describe('verifying the HTML structure', () => {
        const props = {
            ...REQUIRED_PROPS,
            ...OPTIONAL_PROPS,
        };

        it('should expect nodes/dom elements with `display block`', () => {
            const selectInputComponent = mount(<SelectInput {...props} />);
            expect(selectInputComponent).toHaveLength(1);
            expect(selectInputComponent.find('div')).toHaveLength(1);
            expect(selectInputComponent.find('div').hasClass('Input SelectInput block')).toBe(true);

            expect(selectInputComponent.find('div').find('select')).toHaveLength(1);
            expect(
                selectInputComponent
                    .find('div')
                    .find('select')
                    .hasClass(props.className),
            ).toBe(true);

            const listItems = selectInputComponent
                .find('div')
                .find('select')
                .find('option');
            expect(listItems).toHaveLength(3);

            expect(listItems.at(0).props().value).toEqual('apple');
            expect(listItems.at(0).text()).toEqual('Apple');

            expect(listItems.at(1).text()).toEqual('Banana');
            expect(listItems.at(1).props().value).toEqual('banana');

            expect(listItems.at(2).text()).toEqual('Peach');
            expect(listItems.at(2).props().value).toEqual('peach');
        });

        it('should expect nodes/dom elements with `display inline`', () => {
            const selectInputComponent = mount(<SelectInput {...props} display={'inline'} />);
            expect(selectInputComponent).toHaveLength(1);
            expect(selectInputComponent.find('div')).toHaveLength(1);
            expect(selectInputComponent.find('div').hasClass('Input SelectInput inline')).toBe(true);

            expect(selectInputComponent.find('div').find('select')).toHaveLength(1);
            expect(
                selectInputComponent
                    .find('div')
                    .find('select')
                    .hasClass(props.className),
            ).toBe(true);

            expect(
                selectInputComponent
                    .find('div')
                    .find('select')
                    .find('option'),
            ).toHaveLength(3);
            const listItems = selectInputComponent
                .find('div')
                .find('select')
                .find('option');

            expect(listItems.at(0).props().value).toEqual('apple');
            expect(listItems.at(0).text()).toEqual('Apple');

            expect(listItems.at(1).text()).toEqual('Banana');
            expect(listItems.at(1).props().value).toEqual('banana');

            expect(listItems.at(2).text()).toEqual('Peach');
            expect(listItems.at(2).props().value).toEqual('peach');
        });

        it('should expect nodes/dom elements with `label` and `required`', () => {
            const selectInputComponent = mount(<SelectInput {...props} required label={'Select your fruit'} />);
            expect(selectInputComponent.find('div').find('label')).toHaveLength(1);
            expect(
                selectInputComponent
                    .find('div')
                    .find('label')
                    .text(),
            ).toEqual('Select your fruit*');
        });

        it('should expect nodes/dom elements with `error`', () => {
            const selectInputComponent = mount(<SelectInput {...props} inputError={'Alarm! Meltdown Imminent!'} />);
            expect(selectInputComponent.find('div').find('span')).toHaveLength(1);
            expect(
                selectInputComponent
                    .find('div')
                    .find('span')
                    .text(),
            ).toEqual('Alarm! Meltdown Imminent!');
        });

        it('should expect nodes/dom elements with `disabled` value', () => {
            const selectInputComponent = mount(<SelectInput {...props} disabled />);
            expect(selectInputComponent.find('div')).toHaveLength(1);
            expect(selectInputComponent.find('div').hasClass('Input SelectInput block disabled')).toBe(true);
            expect(
                selectInputComponent
                    .find('div')
                    .find('select')
                    .props().disabled,
            ).toBe(true);
        });

        it('should render with `placeholder`', () => {
            const component = mount(<SelectInput {...REQUIRED_PROPS} placeholder="My placeholder" />);

            expect(component.find('.SelectInput__Placeholder')).toHaveLength(1);
            expect(component.find('.SelectInput__Placeholder').text()).toEqual('My placeholder');
        });
    });
});
