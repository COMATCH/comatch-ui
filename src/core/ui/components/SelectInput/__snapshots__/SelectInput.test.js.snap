// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectInput tests should render and check snapshot correctly 1`] = `
<div
  className="Input SelectInput block"
  id={null}
>
  
  <select
    className="SelectInput__input"
    disabled={false}
    name="fruits"
    onBlur={[Function]}
    onChange={[Function]}
  >
    <option
      value="apple"
    >
      Apple
    </option>
    <option
      value="banana"
    >
      Banana
    </option>
    <option
      value="peach"
    >
      Peach
    </option>
  </select>
  
</div>
`;
